docs/root/overview.md

```diff
@@ -1,60 +1,70 @@
-### Phases
-
-One notable addition / change that may get proposed in the near future is the
-introduction of **Phase**. **Phase** would represent a distinct stage of an
-execution, for example a warm-up. It would then be useful to have
-per-phase reporting of latencies as well as counters and latencies.
-
-Concretely, a warm-up phase could be represented by a distinct duration
-termination predicate and a ramping rate limiter. Upon completion, the `hot`
-BenchmarkClient with its associated pool would then be live-transferred to the
-next configured phase, after which execution can continue.
-
-One other reason to have this is that it would enable remote- and/or cli-
-controlled ramping of certain test parameters by associating those to different
-phases. Termination predicates can be leveraged to immediately terminate the
-current phase after injecting a new one, allowing for real-time steering via
-gRPC and/or CLI.
-
-### Streaming parameterization and output stats
-
-Once we have phases, the gRPC service, and perhaps the CLI, would be natural
-candidates to follow up with to allow dynamic phase injection, as well as send
-back reports per phase.
+## Architecture
+
+The following diagram outlines the overall architecture of Nighthawk.
+
+![Nighthawk architecture](images/nighthawk_architecture.png)
+
+Nighthawk exposes its API either via a [command line
+interface](../../source/exe/client_main_entry.cc) (CLI) or as a [gRPC
+server](../../source/exe/service_main_entry.cc) with a [proto
+API](../../api/client/service.proto).
+When running as a CLI, Nighthawk will run a single execution and print out the
+results. When running as a gRPC server, Nighthawk is capable of performing
+multiple executions one after another, returning results to the gRPC client.
+
+In either of the execution modes, the
+[Options](../../source/client/options_impl.h) object abstracts all the
+[options](../../api/client/options.proto) selected by the user and provides them
+to the rest of the Nighthawk architecture.
+
+Nighthawk creates a [Process](../../source/client/process_impl.h)
+instance that generates Envoy bootstrap configuration capable of upstreaming
+requests with the protocol and options selected by the user. This bootstrap is
+used to initialize an Envoy upstream cluster manager.
+
+The **Process** instance creates multiple [Client
+workers](../../source/client/client_worker_impl.h) that get access to the Envoy
+upstream cluster manager and drive the execution.
+See the [next section](#client-worker-architecture) for a detail view into the
+worker architecture.
+
+Once the execution finishes, the [Output
+Collector](../../source/client/output_collector_impl.h)
+provides results collected from individual **Client workers** to the [Output
+formatter](../../source/client/output_formatter_impl.h)
+which represents the results in the format selected by the user.
+
+### Client worker architecture
+
+![Client worker architecture](images/client_worker_architecture.png)
```

Comment:
For this diagram:

1. Can we indicate that the request source is optional? Maybe with a dotted arrow?
2. Can we add in the user here?  I think some arrows would help there, particularly with making clearer what happens to the things you are calling results

Reply:
As discussed offline, request source isn't optional. If we aren't using the plugin or the gRPC request source, we still need an instance of request source to source the requests, in which case Nighthawk uses the built-in one.

Added a user icon to the diagram.

Done in e4079385796871b10bbb9e7b6a0c5dee433af0e8.